### PR TITLE
Remove hard-coded database schema

### PR DESCRIPTION
**Pull request by jmagnusson**
Makes the code more portable by defaulting to not using a separate schema. In response to the discussion in #3.

### CODE REVIEW

**Comment by jmagnusson** [comment]
@kvesteri I've added schema support back in as per your comments in #3. More work than I first imagined :stuck_out_tongue_winking_eye: 

Defaults to using the public schema as it will work out of the box for more people. For a custom schema you'll have to create your own `VersioningManager` instance. To get back the old behavior:

```python
from postgresql_audit import VersioningManager
versioning_manager = VersioningManager(schema_name='audit')
versioning_manager.init(my_base)
```

**Comment by kvesteri** [comment]
Can you make a separate pull request of the tox test environment? Also we should use different continous integration builds for different python environments. This can be done with something like the following:

```
env:
  matrix:
    - TOX_ENV=lint
    - TOX_ENV=py27
    - TOX_ENV=py33
    - TOX_ENV=py34
    - TOX_ENV=pypy
install:
  - pip install tox
script:
  - tox -e $TOX_ENV
```

File: postgresql_audit/base.py [hunk 1]
```diff
@@ -55,32 +51,10 @@ def assign_actor(base, cls, actor_cls):
...
-    )
-    if query not in cached_statements:
-        cached_statements[query] = StatementExecutor(query)
-    listener = (table, 'after_create', cached_statements[query])
-    if not sa.event.contains(*listener):
-        sa.event.listen(*listener)
-
-
-def activity_base(base):
+def activity_base(base, schema='public'):
```
Comment (kvesteri) [inline postgresql_audit/base.py]: Lets leave the default value to ``None`` and use whatever default value SQLAlchemy uses for the schema. Same applies to all functions that have this ``schema`` parameter.

**Comment by jmagnusson** [comment]
> Can you make a separate pull request of the tox test environment? Also we should use different continous integration builds for different python environments. This can be done with something like the following:

Fixed in #9. Will remove from this PR. Btw you don't need to pass in `TOX_ENV` if you use the variable name `TOXENV`, as it's always read by tox.

File: postgresql_audit/base.py [hunk 2]
```diff
@@ -1,16 +1,18 @@
-import os
+import warnings
 from datetime import timedelta
 from weakref import WeakSet
 
 import sqlalchemy as sa
+from jinja2 import Environment, PackageLoader
```
Comment (kvesteri) [inline postgresql_audit/base.py]: I'd like to avoid this. I think PostgreSQL-Audit shouldn't have a strong dependency on Jinja2.

File: postgresql_audit/base.py [hunk 3]
```diff
@@ -168,17 +127,86 @@ def __init__(self, actor_cls=None):
...
         )
+        self.schema_name = schema_name
+        self.table_listener_mapping = self.get_table_listener_mapping()
+        self.table_listeners = list(self.table_listener_mapping.items())
         self.pending_classes = WeakSet()
         self.cached_ddls = {}
 
-    @property
-    def transaction_values(self):
+    def default_actor_id(self):
```
Comment (kvesteri) [inline postgresql_audit/base.py]: This and the method bewlo could be in different commit and pull request?

File: postgresql_audit/base.py [hunk 4]
```diff
@@ -168,17 +127,86 @@ def __init__(self, actor_cls=None):
...
+        if not sa.event.contains(*listener):
+            sa.event.listen(*listener)
+
     def set_activity_values(self, session):
+        dialect = session.bind.engine.dialect
         table = self.activity_cls.__table__
-        if self.values:
-            values = convert_callables(self.transaction_values)
+
+        if not isinstance(dialect, PGDialect):
```
Comment (kvesteri) [inline postgresql_audit/base.py]: This could also be in different commit & PR

File: postgresql_audit/flask.py [hunk 5]
```diff
@@ -6,17 +6,40 @@
...
-        if hasattr(g, 'activity_values'):
+        if context_available() and hasattr(g, 'activity_values'):
             values.update(g.activity_values)
+        if 'client_addr' not in values:
+            values['client_addr'] = self.default_client_addr()
+        if 'actor_id' not in values:
+            values['actor_id'] = self.default_actor_id()
         return values
 
+    def default_actor_id(self):
```
Comment (kvesteri) [inline postgresql_audit/flask.py]: Same here, let's put this refactoring stuff in its own PR

File: tests/test_custom_schema.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    yield
+    base.metadata.drop_all(connection)
+    session.commit()
+
+
+@pytest.mark.usefixtures('cs_activity_cls',
+                         'cs_table_creator')
+class TestCustomSchemaActivityCreation(object):
+
+    def test_cs_insert(self, user, connection, cs_name):
```
Comment (kvesteri) [inline tests/test_custom_schema.py]: I think we could remove the cs prefix from these methods. It is implicitly clear from the class name.

File: tests/test_custom_schema.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    tx = connection.begin()
+    cs_versioning_manager.activity_cls.__table__.create(connection)
+    base.metadata.create_all(connection)
+    tx.commit()
+    yield
+    base.metadata.drop_all(connection)
+    session.commit()
+
+
+@pytest.mark.usefixtures('cs_activity_cls',
```
Comment (kvesteri) [inline tests/test_custom_schema.py]: Let's always use indentation such as

```python
some_very_long_function_name(
    arg1,
    arg2
)
```

Over

```python
some_very_long_function_name(arg1,
                             arg2
)
```

File: postgresql_audit/base.py (hunk 2 above)
Comment (jmagnusson) [inline postgresql_audit/base.py]: How should I go about doing stuff like this in the templates then?

```sql
{% if schema_prefix != '' -%}
REVOKE ALL ON {{ schema_prefix }}activity FROM public;
{%- endif %}
```

File: postgresql_audit/flask.py (hunk 5 above)
Comment (jmagnusson) [inline postgresql_audit/flask.py]: I don't have a good alternative to this now that the tests are run without using the global `versioning_manager` instances. Is it okay that we just keep this in the PR? I'm feeling a bit exhausted by all the rebasing and refactoring 😩

**Comment by jmagnusson** [comment]
I've updated the PR accordingly, except for the parts were I left comments above.

File: tests/test_custom_schema.py (hunk 7 above)
Comment (jmagnusson) [inline tests/test_custom_schema.py]: I think I caught

File: postgresql_audit/base.py (hunk 2 above)
Comment (kvesteri) [inline postgresql_audit/base.py]: All logic like this can be kept on python side.

File: postgresql_audit/flask.py [hunk 8]
```diff
@@ -6,17 +6,40 @@
...
 from flask.globals import _app_ctx_stack, _request_ctx_stack
 
-from .base import VersioningManager
+from .base import VersioningManager as BaseVersioningManager
 
 
-class FlaskVersioningManager(VersioningManager):
-    @property
-    def transaction_values(self):
+class VersioningManager(BaseVersioningManager):
```
Comment (kvesteri) [inline postgresql_audit/flask.py]: Why change the the name? I'd like to keep the name as ``FlaskVersioningManager`` as its more explicit. Also bear in mind that at one point we can support other frameworks as well (for example Bottle or Pyramid).

File: postgresql_audit/flask.py (hunk 5 above)
Comment (kvesteri) [inline postgresql_audit/flask.py]: Fair enough :)

File: postgresql_audit/flask.py (hunk 8 above)
Comment (jmagnusson) [inline postgresql_audit/flask.py]: I was thinking it was more clean as it's in another module. But I can change back if you'd like?

**Comment by kvesteri** [comment]
Other than the Jinja2 dependency, this looks good now! If you can get rid of that, I'll merge this.

**Comment by jmagnusson** [comment]
Done! Replaced with string.Template.

**Comment by kvesteri** [comment]
Thanks for the PR! :train:
